<a href="https://colab.research.google.com/github/Mayar455/Data-Science-/blob/main/intro_to_data_science.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ***INTRO TO DATA SCIENCE ***
*notes from sololearn and codes*
-----------------------------------------

**What is data Science ?**

Data science is a multidisciplinary field that unifies statistics, data analysis, machine learning and their related methods to extract knowledge and insights.

1- No matter the format of the data, the first step in data science is to transform it into arrays of numbers.

# **Numpy**

In [1]:
import numpy as np

heights = [189, 170, 189, 163, 183, 171, 185, 168, 173, 183, 173, 173, 175, 178, 183, 193, 178, 173, 174, 183, 183, 180, 168, 180, 170, 178, 182, 180, 183, 178, 182, 188, 175, 179, 183, 193, 182, 183, 177, 185, 188, 188, 182, 185, 191]

heights_arr = np.array(heights)
print((heights_arr > 188).sum())

5


In [2]:
#An array class in Numpy is called an ndarray or n-dimensional array.
#We can use this to count the number of presidents in heights_arr, use attribute numpy.ndarray.size
#Note that once an array is created in numpy, its size cannot be changed.
#attribute size is the same as len built-in function
heights_arr.size

45

*Size tells us how big the array is, shape tells us the dimension. To get current shape of an array use attribute shape:*

In [3]:
heights_arr.shape

(45,)

In [4]:
ages = [57, 61, 57, 57, 58, 57, 61, 54, 68, 51, 49, 64, 50, 48, 65, 52, 56, 46, 54, 49, 51, 47, 55, 55, 54, 42, 51, 56, 55, 51, 54, 51, 60, 62, 43, 55, 56, 61, 52, 69, 64, 46, 54, 47, 70]
heights_and_ages = heights + ages 
# convert a list to a numpy array
heights_and_ages_arr = np.array(heights_and_ages)
print(heights_and_ages_arr.shape)

(90,)


It would be clearer if we could align height and age for each president and reorganize the data into a 2 by 45 matrix where the first row contains all heights and the second row contains ages.To achieve this, a new array can be created by calling numpy.ndarray.reshape with new dimensions specified in a tuple.

We can reshape an array in multiple ways, as long as the size of the reshaped array matches that of the original. *italicized text* *italicized text*

In [5]:

heights_and_ages_arr.reshape((2,45))

array([[189, 170, 189, 163, 183, 171, 185, 168, 173, 183, 173, 173, 175,
        178, 183, 193, 178, 173, 174, 183, 183, 180, 168, 180, 170, 178,
        182, 180, 183, 178, 182, 188, 175, 179, 183, 193, 182, 183, 177,
        185, 188, 188, 182, 185, 191],
       [ 57,  61,  57,  57,  58,  57,  61,  54,  68,  51,  49,  64,  50,
         48,  65,  52,  56,  46,  54,  49,  51,  47,  55,  55,  54,  42,
         51,  56,  55,  51,  54,  51,  60,  62,  43,  55,  56,  61,  52,
         69,  64,  46,  54,  47,  70]])

In [6]:
# inspect the first three elements from the first row in a 2darray
heights_and_ages_arr = heights_and_ages_arr.reshape((2,45))
heights_and_ages_arr[0, 0:3]

array([189, 170, 189])

In [7]:
#Second column of the 2d array 
heights_and_ages_arr[:, 1]

array([170,  61])

In [8]:
#change the fourth entry in heights_arr to 165
heights_and_ages_arr[0, 3] =165
heights_and_ages_arr

array([[189, 170, 189, 165, 183, 171, 185, 168, 173, 183, 173, 173, 175,
        178, 183, 193, 178, 173, 174, 183, 183, 180, 168, 180, 170, 178,
        182, 180, 183, 178, 182, 188, 175, 179, 183, 193, 182, 183, 177,
        185, 188, 188, 182, 185, 191],
       [ 57,  61,  57,  57,  58,  57,  61,  54,  68,  51,  49,  64,  50,
         48,  65,  52,  56,  46,  54,  49,  51,  47,  55,  55,  54,  42,
         51,  56,  55,  51,  54,  51,  60,  62,  43,  55,  56,  61,  52,
         69,  64,  46,  54,  47,  70]])

a 1d array or a 2d arry can be assigned to a subset of another 2d array, as long as their shapes match.

In [9]:
#update both height and age of the first president with new data, we can supply the data in a list:
heights_and_ages_arr[:, 0] = [190, 58]
heights_and_ages_arr

array([[190, 170, 189, 165, 183, 171, 185, 168, 173, 183, 173, 173, 175,
        178, 183, 193, 178, 173, 174, 183, 183, 180, 168, 180, 170, 178,
        182, 180, 183, 178, 182, 188, 175, 179, 183, 193, 182, 183, 177,
        185, 188, 188, 182, 185, 191],
       [ 58,  61,  57,  57,  58,  57,  61,  54,  68,  51,  49,  64,  50,
         48,  65,  52,  56,  46,  54,  49,  51,  47,  55,  55,  54,  42,
         51,  56,  55,  51,  54,  51,  60,  62,  43,  55,  56,  61,  52,
         69,  64,  46,  54,  47,  70]])

In [10]:
#We can also update data in a subarray with a numpy array as such:
new_record = np.array([[180, 183, 190], [54, 50, 69]])
heights_and_ages_arr[:, 42:] = new_record
heights_and_ages_arr


array([[190, 170, 189, 165, 183, 171, 185, 168, 173, 183, 173, 173, 175,
        178, 183, 193, 178, 173, 174, 183, 183, 180, 168, 180, 170, 178,
        182, 180, 183, 178, 182, 188, 175, 179, 183, 193, 182, 183, 177,
        185, 188, 188, 180, 183, 190],
       [ 58,  61,  57,  57,  58,  57,  61,  54,  68,  51,  49,  64,  50,
         48,  65,  52,  56,  46,  54,  49,  51,  47,  55,  55,  54,  42,
         51,  56,  55,  51,  54,  51,  60,  62,  43,  55,  56,  61,  52,
         69,  64,  46,  54,  50,  69]])

In [ ]:
#nstead of having the ages stored in a list, it could be stored in a 2darray:
heights_arr = np.array(heights)
ages_arr = np.array(ages)

heights_arr = heights_arr.reshape((45,1))
ages_arr = ages_arr.reshape((45,1))

#If we reshape the heights_arr to (45,1), the same as 'ages_arr',
# we can stack them horizontally (by column) to get a 2darray using 'hstack':

height_age_arr = np.hstack((heights_arr, ages_arr))
print(height_age_arr.shape)
print(height_age_arr[:3,])


Similarly, if we want to combine the arrays vertically (by row), we can use **'vstack'.**

In [ ]:
# If we want to concatenate, link together, two arrays along rows,
# then pass 'axis = 1' to achieve the same result as using numpy.hstack; 
# and pass 'axis = 0' if you want to combine arrays vertically.
height_age_arr = np.concatenate((heights_arr, ages_arr), axis=1)

print(height_age_arr.shape)
print(height_age_arr[:3,:])

***NOTE***: You can use **np.hstack** to concatenate arrays ONLY if they have the same **number of rows.**

In [ ]:
# mathematical operations for addition, subtraction, division and power (+, -, /, **) work the same way on arrays.
print(height_age_arr[:,0]*0.0328084)

In [28]:
height_age_arr.sum()



10575

The sum of all heights and ages is 10575. In order to sum all heights and sum all ages separately, we can specify axis=0 to calculate the sum across the rows, that is, it computes the sum for each column, or column sum. On the other hand, to obtain the row sums specify axis=1

In [29]:
height_age_arr.sum(axis=0)

array([8100, 2475])

In [30]:
height_age_arr.sum(axis=1)

array([246, 231, 246, 220, 241, 228, 246, 222, 241, 234, 222, 237, 225,
       226, 248, 245, 234, 219, 228, 232, 234, 227, 223, 235, 224, 220,
       233, 236, 238, 229, 236, 239, 235, 241, 226, 248, 238, 244, 229,
       254, 252, 234, 236, 232, 261])



instead of the entire dataset, we want only tall presidents, that is, those presidents whose height is greater than or equal to 182 cm. We first create a mask, 1darray with boolean values:

In [31]:
mask = height_age_arr[:, 0] >= 182
print(mask.sum())

23


Then pass it to the first axis of `height_age_arr` to filter presidents who don’t meet the criteria:

In [32]:
tall_presidents = height_age_arr[mask, ]
tall_presidents.shape

tall_presidents.shape

(23, 2)

NOTE; **MASKING ** is used to extract, modify, count, or otherwise manipulate values in an array based on some criterion. In our example, the criteria was height of 182cm or taller.

In [33]:
#multiple masking 

mask = (height_age_arr[:, 0]>=182) & (height_age_arr[:,1]<=50)
height_age_arr[mask,]

array([[183,  49],
       [183,  43],
       [188,  46],
       [185,  47]])

# **Pandas**

Suppose we want to inspect the data on Abraham Lincoln in 'height_age_arr' but cannot remember his integer position. Is there a convenient way to access the data by indexing the name of the president like:

**print(height_age_arr['Abraham Lincoln'])**

Unfortunately, we will receive an** error message**.  However, it is possible to do this in **pandas**.

The pandas library is built on top of numpy, meaning a lot of features, methods, and functions are shared.


In [ ]:
# Pandas Series is a one-dimensional labeled array that can
# hold data of any type (integer, string, float, python objects, etc.), 
# similar to a column in an excel spreadsheet. The axis labels are collectively called index.

import pandas as pd
pd.Series([1,2,3], index=['a','b','c'])  # with index

In [ ]:
pd.Series(np.array([1,2,3]), index=['a', 'b', 'c']) #from 1D array

In [ ]:
pd.Series({'a':1, 'b':2, 'c':3}) #from a Dict

data is usually more than one-dimensional, and of different data types; thus Series is not sufficient. **DataFrames** are 2darrays with both row and column labels.

In [40]:
# one way to create a DataFrame from scratch is to pass in a dict. For example,
# this week, we sold 3 bottles of red wine to Adam, 6 to Bob, and 5 to Charles.
# We sold 5 bottles of white wine to Adam, 0 to Bob and 10 to Charles. We can organize
# the data into a DataFrame by creating a dict 'wine_dict' with the number of bottles 
# of each wine type we sold, then pass it along with the customer names as index to create a DataFrame 'sales'.
wine_dict = {
    'red_wine' : [3, 6, 5],
    'white_wine': [5, 0, 10]
}
sales = pd.DataFrame(wine_dict, index = ["adam", "bob", "charles"])
sales

red_wine  white_wine
adam            3           5
bob             6           0
charles         5          10